### DISCUSSION TITLE
Adam optimiser accuracy not changing

### DISCUSSION THREAD

**Discussion by yhl-54** [Part-II-Computational-Physics/.github#83] [Q&A]
Hi, I've tried different values of eta however nothing seems to make my accuracy change. I don't really know where the issue is
Here is my code for reference

def adam(model, n_steps=100):
    '''
    A function that optimises the parameters of the neural network in n_steps steps using the Adam optimiser.
    Return a list of losses and a list of accuracies during training. 
    '''
    
    losses = []
    accuracies = []
    
    b1 = 0.9
    b2 = 0.999
    eps = 1e-8
    lmbda = 1.0
    eta = 5e-2
    
    Nparam = len(model.parameters())
    
    m = np.zeros(Nparam)
    v = np.zeros(Nparam)
    
    for step in range(n_steps):
        
        t = step + 1
    
        # forward
        total_loss, acc = loss(X, y, model)
        losses.append(total_loss.data)
        accuracies.append(acc)

        # backward
        # TO DO
        model.zero_grad()
        total_loss.backward
        
        # update (adam)
        lr = eta * learning_rate(t, n_steps)  #I have also tried this with learning_rate(step, n_steps)
        #TO DO
        for i,p in enumerate(model.parameters()):
            g = p.grad 
            
            #update EMA of g and g**2
            m[i] = b1 * m[i] + (1 - b1) * g
            v[i] = b2 * v[i] + (1 - b2) * (g**2)
            
            #bias correction
            m_hat = m[i] / (1 - b1**t)
            v_hat = v[i] / (1 - b2**t)
            
            #update model parameters
            p.data = p.data - lr * (m_hat / (np.sqrt(v_hat) + eps) + lmbda * p.data) 
            
        if t >= n_steps:
            break

        if step % 10 == 0 or step == n_steps-1:
            print(f"step {step} loss {total_loss.data}, accuracy {acc*100}%")
    
    return losses, accuracies

And the code for my losses
def loss(Xdat,ydat,model):
    """Returns the total loss (C(theta) + regularisation) and the accuracy of the neural network."""
    
    inputs = [list(map(Value, xrow)) for xrow in Xdat]
    scores = list(map(model, inputs))
    
    #TO DO: implement data_loss
    SVM = [(1 + -yi*scorei).relu() for yi, scorei in zip(ydat, scores)]
    data_loss = (1/len(SVM)) * np.sum(SVM)
                       
    alpha = 1e-4
    reg_loss = alpha * sum((p*p for p in model.parameters()))

    total_loss = data_loss + reg_loss #loss function + regularisation

    accuracy = [(yi > 0) == (scorei.data > 0) for yi, scorei in zip(ydat, scores)]
    
    return total_loss,  sum(accuracy) / len(accuracy)

**Comment by ceciliegl** [comment]
Hi @yhl-54 ,

I cannot see anything wrong with your code. Have you remembered to re-initialise your neural network before you train it?

**Reply by yhl-54** [reply]
Hi! I'm not sure what that means, what would I need to do to re-initialise?

**Reply by ceciliegl** [reply]
It means that you should construct a new neural network through initialise() before you start training to reset the weights and biases of the neural network:

modeladam = initialise(n_neurons = 16)
lossadam, accadam = adam(modeladam)

**Reply by yhl-54** [reply]
Yes in that case, I am re-initialising prior to running it, i can get it working for gd but somehow not for Adam nor lion

**Reply by ceciliegl** [reply]
Hmm, have you tried to restart the kernel?

**Reply by wignalex** [reply]
had a similar problem, fixed it by doing

`import micrograd`

at the top
